controllers/resource_reconciler.go

```diff
@@ -142,57 +97,121 @@ func (r *ResourceReconciler) Pause(ctx context.Context) error {
-
-		log := log.WithValues("resource", klog.KObj(spec), "resourceVersion", gvk.GroupVersion(), "resourceKind", gvk.Kind)
-
-		log.V(4).Info("pausing resource")
+		gvk := spec.GroupVersionKind()
+		log.V(4).Info("pausing resource", "resource", klog.KObj(spec), "resourceVersion", gvk.GroupVersion(), "resourceKind", gvk.Kind)
 		err := r.Patch(ctx, spec, client.Apply, client.FieldOwner("capz-manager"))
-		if err != nil {
+		if client.IgnoreNotFound(err) != nil {
 			return fmt.Errorf("failed to patch resource: %w", err)
 		}
 	}
 
 	return nil
 }
 
-// Delete deletes the specified resources.
-func (r *ResourceReconciler) Delete(ctx context.Context) error {
-	ctx, log, done := tele.StartSpanWithLogger(ctx, "controllers.ResourceReconciler.Delete")
+func (r *ResourceReconciler) reconcile(ctx context.Context) (bool, error) {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "controllers.ResourceReconciler.reconcile")
 	defer done()
-	log.V(4).Info("deleting resources")
 
 	var newResourceStatuses []infrav1alpha.ResourceStatus
 
-	for _, spec := range r.owner.GetResourceStatuses() {
-		newStatus, err := r.deleteResource(ctx, spec.Resource)
+	ownedKindsValue := r.owner.GetAnnotations()[ownedKindsAnnotation]
+	ownedKinds, err := parseOwnedKinds(ownedKindsValue)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse %s annotation: %s", ownedKindsAnnotation, ownedKindsValue)
+	}
+
+	ownedObjs, err := r.ownedObjs(ctx, ownedKinds)
+	if err != nil {
+		return false, fmt.Errorf("failed to get owned objects: %w", err)
+	}
+
+	unobservedTypeResources, observedTypeResources, deletedResources := partitionResources(ownedKinds, r.resources, ownedObjs)
+
+	// Newly-defined types in the CAPZ spec are first recorded in the annotation without performing a
+	// patch that would create resources of that type. CAPZ only patches resources whose kinds have
+	// already been recorded to ensure no resources are orphaned.
+	for _, spec := range unobservedTypeResources {
+		newResourceStatuses = append(newResourceStatuses, infrav1alpha.ResourceStatus{
+			Resource: statusResource(spec),
+			Ready:    false,
+		})
+	}
+
+	for _, spec := range observedTypeResources {
+		spec.SetNamespace(r.owner.GetNamespace())
+
+		if err := controllerutil.SetControllerReference(r.owner, spec, r.Scheme()); err != nil {
+			return false, fmt.Errorf("failed to set owner reference: %w", err)
+		}
+
+		toWatch := meta.AsPartialObjectMetadata(spec)
+		toWatch.APIVersion = spec.GetAPIVersion()
+		toWatch.Kind = spec.GetKind()
+		if err := r.watcher.Watch(log, toWatch, handler.EnqueueRequestForOwner(r.Client.Scheme(), r.Client.RESTMapper(), r.owner)); err != nil {
+			return false, fmt.Errorf("failed to watch resource: %w", err)
+		}
+
+		gvk := spec.GroupVersionKind()
+		log.V(4).Info("applying resource", "resource", klog.KObj(spec), "resourceVersion", gvk.GroupVersion(), "resourceKind", gvk.Kind)
+		err := r.Patch(ctx, spec, client.Apply, client.FieldOwner("capz-manager"), client.ForceOwnership)
```

Comment:
For my understanding, how is the controller's GUID set upon creation ? Is `client.ForceOwnership` enforcing the controller's GUID ?